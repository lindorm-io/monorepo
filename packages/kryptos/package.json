{
  "name": "@lindorm/kryptos",
  "version": "0.3.5",
  "license": "AGPL-3.0-or-later",
  "author": "Jonn Nilsson",
  "repository": {
    "url": "https://github.com/lindorm-io/monorepo",
    "directory": "packages/kryptos"
  },
  "bugs": "https://github.com/lindorm-io/monorepo/issues",
  "publishConfig": {
    "access": "public"
  },
  "main": "dist/index.js",
  "typings": "dist/index.d.ts",
  "scripts": {
    "build": "rimraf dist && tsc -b ./tsconfig.build.json",
    "example": "ts-node example",
    "integration": "compd --file docker-compose.yml jest --config jest.config.integration.js --watch --",
    "test:ci": "npm run test:unit",
    "test:integration": "jest --config jest.config.integration.js --",
    "test:unit": "jest --config jest.config.js --",
    "test": "jest --watch --",
    "typecheck:ci": "tsc",
    "typecheck": "tsc --watch",
    "update": "ncu -u"
  },
  "dependencies": {
    "@lindorm/b64": "^0.1.4",
    "@lindorm/case": "^0.1.5",
    "@lindorm/date": "^0.1.5",
    "@lindorm/errors": "^0.1.6",
    "@lindorm/is": "^0.1.6",
    "@lindorm/random": "^0.1.2",
    "@lindorm/utils": "^0.3.0"
  },
  "devDependencies": {
    "@lindorm/types": "^0.2.0"
  }
}
