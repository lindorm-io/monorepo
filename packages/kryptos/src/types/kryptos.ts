import { Optional } from "@lindorm/types";
import { EcGenerate } from "./ec";
import { KryptosEncryption } from "./encryption";
import { OctGenerate } from "./oct";
import { OkpGenerate } from "./okp";
import { RsaGenerate, RsaModulus } from "./rsa";
import {
  KryptosAlgorithm,
  KryptosCurve,
  KryptosOperation,
  KryptosType,
  KryptosUse,
} from "./types";

export type KryptosAttributes = {
  id: string;
  algorithm: KryptosAlgorithm;
  createdAt: Date;
  curve: KryptosCurve | undefined;
  encryption: KryptosEncryption | undefined;
  expiresAt: Date;
  isExternal: boolean;
  issuer: string | undefined;
  jwksUri: string | undefined;
  notBefore: Date;
  operations: Array<KryptosOperation>;
  ownerId: string | undefined;
  type: KryptosType;
  updatedAt: Date;
  use: KryptosUse;
};

export type KryptosMetadata = {
  expiresIn: number;
  hasPrivateKey: boolean;
  hasPublicKey: boolean;
  isActive: boolean;
  isExpired: boolean;
  isUsable: boolean;
  modulus: RsaModulus | undefined;
};

export type KryptosLike = Partial<KryptosAttributes>;

type StdOptions = Optional<
  KryptosAttributes,
  | "id"
  | "createdAt"
  | "curve"
  | "encryption"
  | "expiresAt"
  | "isExternal"
  | "issuer"
  | "jwksUri"
  | "notBefore"
  | "operations"
  | "ownerId"
  | "updatedAt"
>;

export type KryptosKeys = {
  privateKey?: Buffer;
  publicKey?: Buffer;
};

export type KryptosOptions = StdOptions & KryptosKeys;

export type GenerateKryptosConfig = EcGenerate | OctGenerate | OkpGenerate | RsaGenerate;

export type GenerateKryptosOptions = Pick<
  StdOptions,
  | "encryption"
  | "expiresAt"
  | "issuer"
  | "jwksUri"
  | "notBefore"
  | "operations"
  | "ownerId"
>;
