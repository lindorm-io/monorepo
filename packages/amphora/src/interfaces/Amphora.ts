import { IKryptos } from "@lindorm/kryptos";
import { AmphoraConfig, AmphoraJwks, AmphoraQuery } from "../types";

export interface IAmphora {
  config: Array<AmphoraConfig>;
  jwks: AmphoraJwks;
  vault: Array<IKryptos>;

  add(kryptos: Array<IKryptos> | IKryptos): void;
  filter(query: AmphoraQuery): Promise<Array<IKryptos>>;
  find(query: AmphoraQuery): Promise<IKryptos>;
  refresh(): Promise<void>;
  setup(): Promise<void>;
}
