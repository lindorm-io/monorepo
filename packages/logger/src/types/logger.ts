import { Dict } from "@lindorm/types";
import { FilterCallback, LogContext, LogSession } from "./types";

export interface Logger {
  child(): Logger;
  child(context: LogContext): Logger;
  child(session: LogSession): Logger;
  child(context: LogContext, session: LogSession): Logger;

  context(context: LogContext): void;
  filter(path: string, callback?: FilterCallback): void;
  session(session: LogSession): void;

  error(error: Error): void;
  error(message: string, details?: Error | Dict): void;

  warn(message: string, details?: Dict): void;
  info(message: string, details?: Dict): void;
  verbose(message: string, details?: Dict): void;
  debug(message: string, details?: Dict): void;
  silly(message: string, details?: Dict): void;
}
