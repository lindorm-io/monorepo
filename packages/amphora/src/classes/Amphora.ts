import { Conduit, conduitChangeResponseDataMiddleware } from "@lindorm/conduit";
import { isArray, isString, isUrlLike } from "@lindorm/is";
import { IKryptos, KryptosKit, LindormJwk } from "@lindorm/kryptos";
import { ILogger } from "@lindorm/logger";
import { OpenIdConfigurationResponse, OpenIdJwksResponse } from "@lindorm/types";
import { Predicated } from "@lindorm/utils";
import { AmphoraError } from "../errors";
import { IAmphora } from "../interfaces";
import {
  AmphoraConfig,
  AmphoraExternalOption,
  AmphoraJwks,
  AmphoraOptions,
  AmphoraQuery,
} from "../types";

export class Amphora implements IAmphora {
  public readonly issuer: string | null;

  private readonly conduit: Conduit;
  private readonly external: Array<AmphoraExternalOption>;
  private readonly logger: ILogger;

  private readonly _config: Array<AmphoraConfig>;
  private _jwks: Array<LindormJwk>;
  private _setup: boolean;
  private _vault: Array<IKryptos>;

  public constructor(options: AmphoraOptions) {
    this.logger = options.logger.child(["Amphora"]);

    this.conduit = new Conduit({
      alias: "Amphora",
      logger: this.logger,
      middleware: [conduitChangeResponseDataMiddleware()],
      retryOptions: {
        maxAttempts: 10,
      },
    });

    this._config = [];
    this.external = options.external ?? [];
    this.issuer = options.issuer ?? null;
    this._jwks = [];
    this._setup = false;
    this._vault = [];
  }

  // public getters

  public get config(): Array<AmphoraConfig> {
    return this._config;
  }

  public get jwks(): AmphoraJwks {
    if (!this.issuer) {
      throw new AmphoraError("Issuer is required to get JWKS");
    }

    return { keys: this._jwks };
  }

  public get vault(): Array<IKryptos> {
    return this._vault;
  }

  // public methods

  public add(kryptos: Array<IKryptos> | IKryptos): void {
    const array = isArray(kryptos) ? kryptos : [kryptos];

    for (const item of array) {
      if (!item.id) {
        throw new AmphoraError("Id is required when adding Kryptos");
      }

      if (!item.issuer && this.issuer) {
        item.issuer = this.issuer;
      }

      if (!item.issuer) {
        throw new AmphoraError("Issuer is required when adding Kryptos");
      }

      if (item.isExpired) {
        throw new AmphoraError("Kryptos is expired");
      }

      this._vault = this._vault.filter((i) => i.id !== item.id).concat(item);
    }

    this.refreshJwks();
  }

  public async filter(query: AmphoraQuery): Promise<Array<IKryptos>> {
    const filtered = this.filteredKeys(query);
    if (filtered.length) return filtered;

    if (!isString(query.issuer)) {
      throw new AmphoraError("Unable to find Kryptos without issuer");
    }

    const config = await this.issuerConfig(query.issuer);
    await this.refreshExternal(config);

    return this.filteredKeys(query);
  }

  public async find(query: AmphoraQuery): Promise<IKryptos> {
    const [key] = await this.filter(query);
    if (key) return key;

    throw new AmphoraError("Kryptos not found using query");
  }

  public async refresh(): Promise<void> {
    this.logger.silly("Refreshing vault");

    for (const config of this._config) {
      await this.refreshExternal(config);
    }
  }

  public async setup(): Promise<void> {
    if (this._setup) return;

    await this.loadExternalConfig();
    await this.refresh();

    this._setup = true;
  }

  public canEncrypt(): boolean {
    return (
      this.filteredKeys({ $or: [{ operations: ["encrypt"] }, { use: "enc" }] }).length > 0
    );
  }

  public canDecrypt(): boolean {
    return (
      this.filteredKeys({ $or: [{ operations: ["decrypt"] }, { use: "enc" }] }).length > 0
    );
  }

  public canSign(): boolean {
    return (
      this.filteredKeys({ $or: [{ operations: ["sign"] }, { use: "sig" }] }).length > 0
    );
  }

  public canVerify(): boolean {
    return (
      this.filteredKeys({ $or: [{ operations: ["verify"] }, { use: "sig" }] }).length > 0
    );
  }

  // private methods

  private async addExternalConfig(options: AmphoraExternalOption): Promise<void> {
    this.logger.silly("Adding external config", { options });

    if (isUrlLike(options.openIdConfigurationUri)) {
      const { data } = await this.conduit.get<OpenIdConfigurationResponse>(
        options.openIdConfigurationUri,
      );

      this._config.push(data);

      return;
    }

    if (isString(options.issuer) && isUrlLike(options.jwksUri)) {
      this._config.push({ issuer: options.issuer, jwksUri: options.jwksUri });

      return;
    }

    throw new AmphoraError("Invalid issuer options");
  }

  private filteredKeys(query: AmphoraQuery): Array<IKryptos> {
    const vault = this._vault.filter((i) => i.isActive);

    return Predicated.filter<IKryptos>(vault, query).sort(
      (a, b) => b.createdAt.getTime() - a.createdAt.getTime(),
    );
  }

  private async getExternalJwks(issuer: string): Promise<Array<IKryptos>> {
    this.logger.silly("Finding External JWKS", { issuer });

    const config = await this.issuerConfig(issuer);

    const {
      data: { keys },
    } = await this.conduit.get<OpenIdJwksResponse>(config.jwksUri);

    const result: Array<IKryptos> = [];

    for (const jwk of keys) {
      const iss = jwk.iss ?? config.issuer;
      const jku = jwk.jku ?? config.jwksUri;

      const kryptos = KryptosKit.from.jwk({ ...jwk, iss, jku });

      if (kryptos.isExpired) continue;

      this.logger.silly("Adding Kryptos from external source", { kryptos });

      result.push(kryptos);
    }

    return result;
  }

  private async issuerConfig(issuer: string): Promise<AmphoraConfig> {
    if (this.external.length && !this._config.length) {
      await this.loadExternalConfig();
    }

    const config = this._config.find((c) => c.issuer === issuer);

    if (!config) {
      throw new AmphoraError("Issuer not found in config");
    }

    return config;
  }

  private async loadExternalConfig(): Promise<void> {
    this.logger.silly("Loading external config");

    for (const options of this.external) {
      await this.addExternalConfig(options);
    }
  }

  private async refreshExternal(config: AmphoraConfig): Promise<void> {
    const keys = await this.getExternalJwks(config.issuer);

    this._vault = this._vault.filter((i) => i.issuer !== config.issuer).concat(keys);
  }

  private refreshJwks(): void {
    if (this.issuer === null) return;

    this.logger.silly("Refreshing JWKS");

    this._jwks = this._vault
      .filter((i) => i.isActive)
      .filter((i) => !i.isExternal)
      .filter((i) => !i.ownerId)
      .filter((i) => i.hasPublicKey)
      .filter((i) => i.issuer === this.issuer)
      .sort((a, b) => b.createdAt.getTime() - a.createdAt.getTime())
      .map((i) => i.toJWK("public"));
  }
}
