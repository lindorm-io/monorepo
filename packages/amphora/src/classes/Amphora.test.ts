import { KryptosKit } from "@lindorm/kryptos";
import { createMockLogger } from "@lindorm/logger";
import MockDate from "mockdate";
import nock from "nock";
import {
  OPEN_ID_CONFIGURATION_RESPONSE,
  OPEN_ID_JWKS_RESPONSE,
} from "../__fixtures__/auth0";
import {
  TEST_EC_KEY_SIG,
  TEST_OCT_KEY_ENC,
  TEST_OCT_KEY_SIG,
  TEST_OKP_KEY_ENC,
  TEST_OKP_KEY_SIG,
  TEST_RSA_KEY_SIG,
} from "../__fixtures__/keys";
import { Amphora } from "./Amphora";

const MockedDate = new Date("2024-01-01T08:00:00.000Z");
MockDate.set(MockedDate);

describe("Amphora", () => {
  const issuer = "https://test.lindorm.io/";

  let amphora: Amphora;

  beforeEach(() => {
    amphora = new Amphora({ issuer, logger: createMockLogger() });
  });

  describe("add", () => {
    test("should add key to vault", () => {
      amphora.add(TEST_EC_KEY_SIG);

      expect(amphora.vault).toEqual([TEST_EC_KEY_SIG]);
    });

    test("should add multiple keys to vault", () => {
      amphora.add([
        TEST_EC_KEY_SIG,
        TEST_OCT_KEY_SIG,
        TEST_OKP_KEY_SIG,
        TEST_RSA_KEY_SIG,
      ]);

      expect(amphora.vault).toEqual([
        TEST_EC_KEY_SIG,
        TEST_OCT_KEY_SIG,
        TEST_OKP_KEY_SIG,
        TEST_RSA_KEY_SIG,
      ]);
    });

    test("should only keep one copy of each key id", () => {
      amphora.add([TEST_EC_KEY_SIG, TEST_EC_KEY_SIG, TEST_EC_KEY_SIG, TEST_EC_KEY_SIG]);

      expect(amphora.vault).toEqual([TEST_EC_KEY_SIG]);
    });

    test("should update jwks when adding key", () => {
      amphora.add([
        TEST_EC_KEY_SIG,
        TEST_OCT_KEY_SIG,
        TEST_OKP_KEY_SIG,
        TEST_RSA_KEY_SIG,
      ]);

      expect(amphora.jwks).toMatchSnapshot();
    });
  });

  describe("env", () => {
    test("should add keys to vault from env", () => {
      amphora.env([
        "kryptos:eyJlbmMiOiJBMTkyR0NNIiwiaWF0IjoxNzQ0NzA0MjYzLCJrZXlfb3BzIjpbImRlcml2ZUtleSJdLCJuYmYiOjE3NDQ3MDQyNjMsInB1cnBvc2UiOiJ0ZXN0IiwidWF0IjoxNzQ0NzA0MjYzLCJjcnYiOiJQLTM4NCIsIngiOiJGMTgyVlNMMURyRll5b19feVJ3eXlvS3JtT08wVEU0MktxT0pOQk1CNlgxSlFYbGV1MTVqYVpsN3dHdG5XcmxUIiwieSI6IlM3bElSZG45dlh5QnF4S0FSUTZzampLcXlCekt1T3VJM1BYcExlUEZ3bmpXNDduWEVVN2hDMzNydmF5ZzVZbVkiLCJkIjoiVzlRNmZMc2J2NkN0dk1zWUUyOTJha2VqeUlZeHFUY1BGSTQzUE9Fd1dpeVRrMFhhelk4NEREQnpHZlNVNEhmOCIsImtpZCI6IjE2NmM2YWI2LWRmOWYtNGZkYS1hYWI4LTkyMTM5ZWY2NDc5MiIsImFsZyI6IkVDREgtRVMrQTE5MktXIiwidXNlIjoiZW5jIiwia3R5IjoiRUMifQ",
        "kryptos:eyJpYXQiOjE3NDQ3MDQyMjgsImtleV9vcHMiOlsic2lnbiIsInZlcmlmeSJdLCJuYmYiOjE3NDQ3MDQyMjgsInB1cnBvc2UiOiJ0ZXN0IiwidWF0IjoxNzQ0NzA0MjI4LCJjcnYiOiJFZDI1NTE5IiwieCI6IlBqeDJjSWRtS0lkdGh5V2ZEakxjTnlKOWt6RW9ObnlWWjZCckZVZWUxc2ciLCJkIjoiMzhFS1ZrRjZBaGM0RWFUNm9XcWlGajdzejZ3czdjLXk2ZjgycHAzNHNFZyIsImtpZCI6IjM4MTQ0NTdmLTI2OGItNGQyMi1hNjQ0LTZhZTY5YjdjNzRiMSIsImFsZyI6IkVkRFNBIiwidXNlIjoic2lnIiwia3R5IjoiT0tQIn0",
      ]);

      expect(amphora.vault).toEqual([
        expect.objectContaining({
          id: "166c6ab6-df9f-4fda-aab8-92139ef64792",
          type: "EC",
        }),
        expect.objectContaining({
          id: "3814457f-268b-4d22-a644-6ae69b7c74b1",
          type: "OKP",
        }),
      ]);
    });
  });

  describe("filter", () => {
    test("should filter kryptos by active", async () => {
      const key = KryptosKit.clone(TEST_EC_KEY_SIG, {
        notBefore: new Date("2099-01-01T00:00:00.000Z"),
      });

      amphora.add([key, TEST_OCT_KEY_SIG]);

      await expect(amphora.filter({ issuer })).resolves.toEqual([TEST_OCT_KEY_SIG]);
    });

    test("should filter kryptos by issuer", async () => {
      const kryptos = KryptosKit.generate.sig.oct({
        algorithm: "HS256",
        issuer: "https://other.lindorm.io/",
      });

      amphora.add([kryptos, TEST_OCT_KEY_SIG]);

      await expect(amphora.filter({ issuer })).resolves.toEqual([TEST_OCT_KEY_SIG]);
    });

    test("should filter kryptos and sort them by creation date", async () => {
      amphora.add([
        TEST_EC_KEY_SIG,
        TEST_OCT_KEY_SIG,
        TEST_OKP_KEY_SIG,
        TEST_RSA_KEY_SIG,
      ]);

      await expect(amphora.filter({ issuer, hasPrivateKey: true })).resolves.toEqual([
        TEST_RSA_KEY_SIG,
        TEST_OKP_KEY_SIG,
        TEST_OCT_KEY_SIG,
        TEST_EC_KEY_SIG,
      ]);
    });

    test("should filter kryptos in vault using the private query", async () => {
      const { privateKey, ...der } = TEST_OKP_KEY_SIG.export("der");
      const key = KryptosKit.from.der({ issuer, ...der });

      amphora.add([TEST_EC_KEY_SIG, key]);

      await expect(amphora.filter({ issuer, hasPrivateKey: true })).resolves.toEqual([
        TEST_EC_KEY_SIG,
      ]);
    });

    test("should filter kryptos in vault using the public query", async () => {
      amphora.add([TEST_EC_KEY_SIG, TEST_OCT_KEY_SIG]);

      await expect(amphora.filter({ issuer, hasPublicKey: true })).resolves.toEqual([
        TEST_EC_KEY_SIG,
      ]);
    });

    test("should filter kryptos in vault using the operation query", async () => {
      amphora.add([TEST_EC_KEY_SIG, TEST_OCT_KEY_SIG, TEST_OCT_KEY_ENC]);

      await expect(
        amphora.filter({ issuer, operations: ["deriveKey"] }),
      ).resolves.toEqual([TEST_OCT_KEY_ENC]);
    });

    test("should filter kryptos in vault using the type query", async () => {
      amphora.add([TEST_EC_KEY_SIG, TEST_OCT_KEY_SIG]);

      await expect(amphora.filter({ issuer, type: "oct" })).resolves.toEqual([
        TEST_OCT_KEY_SIG,
      ]);
    });

    test("should filter kryptos in vault using the use query", async () => {
      amphora.add([TEST_EC_KEY_SIG, TEST_OCT_KEY_ENC]);

      await expect(amphora.filter({ issuer, use: "sig" })).resolves.toEqual([
        TEST_EC_KEY_SIG,
      ]);
    });

    test("should filter kryptos in vault synchronously", async () => {
      amphora.add(TEST_EC_KEY_SIG);

      expect(amphora.filterSync({ issuer, id: TEST_EC_KEY_SIG.id })).toEqual([
        TEST_EC_KEY_SIG,
      ]);
    });
  });

  describe("find", () => {
    test("should find kryptos in vault using id", async () => {
      amphora.add(TEST_EC_KEY_SIG);

      await expect(amphora.find({ issuer, id: TEST_EC_KEY_SIG.id })).resolves.toEqual(
        TEST_EC_KEY_SIG,
      );
    });

    test("should find kryptos in vault synchronously", async () => {
      amphora.add(TEST_EC_KEY_SIG);

      expect(amphora.findSync({ issuer, id: TEST_EC_KEY_SIG.id })).toEqual(
        TEST_EC_KEY_SIG,
      );
    });
  });

  describe("can", () => {
    test("should return true for canEncrypt", () => {
      amphora.add(TEST_OCT_KEY_ENC);

      expect(amphora.canEncrypt()).toBe(true);
    });

    test("should return false for canEncrypt", () => {
      expect(amphora.canEncrypt()).toBe(false);
    });

    test("should return true for canDecrypt", () => {
      amphora.add(TEST_OCT_KEY_ENC);

      expect(amphora.canDecrypt()).toBe(true);
    });

    test("should return false for canDecrypt", () => {
      expect(amphora.canDecrypt()).toBe(false);
    });

    test("should return true for canSign", () => {
      amphora.add(TEST_RSA_KEY_SIG);

      expect(amphora.canSign()).toBe(true);
    });

    test("should return false for canSign", () => {
      expect(amphora.canSign()).toBe(false);
    });

    test("should return true for canVerify", () => {
      amphora.add(TEST_RSA_KEY_SIG);

      expect(amphora.canVerify()).toBe(true);
    });

    test("should return false for canVerify", () => {
      expect(amphora.canVerify()).toBe(false);
    });
  });

  describe("external config", () => {
    test("should add external config and find jwks", async () => {
      nock("https://lindorm.eu.auth0.com")
        .get("/.well-known/openid-configuration")
        .times(1)
        .reply(200, OPEN_ID_CONFIGURATION_RESPONSE);

      nock("https://lindorm.eu.auth0.com")
        .get("/.well-known/jwks.json")
        .times(1)
        .reply(200, OPEN_ID_JWKS_RESPONSE);

      nock("https://lindorm.jp.auth0.com")
        .get("/.well-known/openid-configuration")
        .times(1)
        .reply(200, {
          ...OPEN_ID_CONFIGURATION_RESPONSE,
          issuer: "https://lindorm.jp.auth0.io/",
          jwksUri: "https://lindorm.jp.auth0.com/.well-known/jwks.json",
        });

      nock("https://lindorm.jp.auth0.com")
        .get("/.well-known/jwks.json")
        .times(1)
        .reply(200, { keys: [TEST_OKP_KEY_ENC.toJWK()] });

      nock("https://external.lindorm.io")
        .get("/.well-known/jwks.json")
        .times(1)
        .reply(200, { keys: [TEST_EC_KEY_SIG.toJWK("private")] });

      amphora = new Amphora({
        issuer,
        logger: createMockLogger(),
        external: [
          {
            issuer: "https://external.lindorm.io/",
            jwksUri: "https://external.lindorm.io/.well-known/jwks.json",
          },
          {
            openIdConfigurationUri:
              "https://lindorm.eu.auth0.com/.well-known/openid-configuration",
          },
          {
            issuer: "https://lindorm.jp.auth0.com/",
          },
        ],
      });

      expect(amphora.config).toEqual([]);

      await amphora.setup();

      expect(amphora.config).toMatchSnapshot();

      expect(amphora.vault).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            id: expect.any(String),
            type: "EC",
          }),
          expect.objectContaining({
            id: expect.any(String),
            type: "OKP",
          }),
          expect.objectContaining({ id: "iPy9pgzr7cFw1kTuiClWE", type: "RSA" }),
          expect.objectContaining({ id: "IjICkHcf-qq8_stUQ00IN", type: "RSA" }),
        ]),
      );
    });

    test("should add use external config when vault is unable to find key", async () => {
      nock("https://external.lindorm.io")
        .get("/.well-known/jwks.json")
        .times(1)
        .reply(200, { keys: [TEST_EC_KEY_SIG.toJWK("private")] });

      amphora = new Amphora({
        issuer,
        logger: createMockLogger(),
        external: [
          {
            issuer,
            jwksUri: "https://external.lindorm.io/.well-known/jwks.json",
          },
        ],
      });

      await expect(amphora.find({ issuer, id: TEST_EC_KEY_SIG.id })).resolves.toEqual(
        expect.objectContaining({ id: TEST_EC_KEY_SIG.id }),
      );
    });
  });
});
