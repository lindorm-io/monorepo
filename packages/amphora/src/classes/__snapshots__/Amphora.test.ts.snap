// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Amphora add should update jwks when adding key 1`] = `
{
  "keys": [
    {
      "alg": "RS512",
      "e": "AQAB",
      "exp": 1717200000,
      "iat": 1704067560,
      "iss": "https://test.lindorm.io/",
      "jku": "https://test.lindorm.io/.well-known/jwks.json",
      "key_ops": [
        "sign",
        "verify",
      ],
      "kid": "aaac22b3-2253-5598-8e0c-1733fc748122",
      "kty": "RSA",
      "n": "7cxLHoxaHdYU8EX2ij5m42XIQYWBUxKMOfVbV2OX0nNHEdojXfRfONvrAHwsvRqcdN44yv1E1be18qs5HRB67D3qLeRjXcjegdjWvQj66IFVZIxQHNJ-LXDbAYs7gez8RYbBObnDzfrqy9vIPZet1tbqvlxv9aTR38x3ORJ4Y6Ym9StbKOUIo7BLIvoc1M5uNUIEnFKKZhbmVR423-F50w_E0618KExZgeur_8cMsSWZxyl0oeSI-rIodjfVJbshT1yy98zKRj-mMnyPKgHQARr9SwDA5Xlt9L2fVZ72xgwaPOBUgFoDTPQsqQ3Rkj7nYUSqwQVEL0-oH2K7U3OeGeKyIQMNzRdSfYMw3gSc5dtwqX64oK33lN2jzTKa0clSZjqJ4jzUL1YrXMddF07dMQqjN2mGjeZVy5LaO4B58Y5l-kO01KdwA-_LcDhkNgbKiZMZVtNPlVkR3nCvInt5cqPOqMOcLW3oA1wmh7zauWqorklpSSs75zw8ETtzATSJv8h7kUG9Bl1QQgdBkGqKb4ZG-r5Bj6pwJrHVmgPSMowP8P9G13iQLFdKX714A3DMKDc03TJvs2yVK81YUCtO5aRdwb0q0YcFqn9BHI3cmO-kOhXquMakPQ3zW_FBSnIeBhJMObCQOmYEuZxrifMcCQTY7j5jsP962e7ZPn53STc",
      "nbf": 1672534800,
      "uat": 1704095940,
      "use": "sig",
    },
    {
      "alg": "EdDSA",
      "crv": "Ed25519",
      "exp": 1717200000,
      "iat": 1704067500,
      "iss": "https://test.lindorm.io/",
      "jku": "https://test.lindorm.io/.well-known/jwks.json",
      "key_ops": [
        "sign",
        "verify",
      ],
      "kid": "2fa52a91-7f63-5731-a55d-30d36350c642",
      "kty": "OKP",
      "nbf": 1672534800,
      "uat": 1704095940,
      "use": "sig",
      "x": "GRCwCA6lChosFGMQwxGiHCdzblfvCz0FNiRtTnm1qqc",
    },
    {
      "alg": "ES512",
      "crv": "P-521",
      "exp": 1717200000,
      "iat": 1704067260,
      "iss": "https://test.lindorm.io/",
      "jku": "https://test.lindorm.io/.well-known/jwks.json",
      "key_ops": [
        "sign",
        "verify",
      ],
      "kid": "b9e7bb4d-d332-55d2-9b33-f990ff7db4c7",
      "kty": "EC",
      "nbf": 1672534800,
      "uat": 1704095940,
      "use": "sig",
      "x": "AI3Y8hWlFN19cuY3GSDwGTabmd62piw1bYt5-GSWvaN8mJVQlQM-hmgZ9psbMcSCwqM_YdFs-wQj-00PfHg31VbR",
      "y": "AAgPZRhdrsY3SLMdYgVF2NVi5Hn7SPLAW136j83Y852u-VXzG6Z1Ur5OJI8Qh18qgv8i7R7sgpOwmyP5SYTysMmf",
    },
  ],
}
`;

exports[`Amphora external config should add external config and find jwks 1`] = `
[
  {
    "issuer": "https://external.lindorm.io/",
    "jwksUri": "https://external.lindorm.io/.well-known/jwks.json",
  },
  {
    "authorizationEndpoint": "https://lindorm.eu.auth0.com/authorize",
    "claimsSupported": [
      "aud",
      "auth_time",
      "created_at",
      "email",
      "email_verified",
      "exp",
      "family_name",
      "given_name",
      "iat",
      "identities",
      "iss",
      "name",
      "nickname",
      "phone_number",
      "picture",
      "sub",
    ],
    "codeChallengeMethodsSupported": [
      "S256",
      "plain",
    ],
    "deviceAuthorizationEndpoint": "https://lindorm.eu.auth0.com/oauth/device/code",
    "idTokenSigningAlgValuesSupported": [
      "HS256",
      "RS256",
      "PS256",
    ],
    "issuer": "https://lindorm.eu.auth0.com/",
    "jwksUri": "https://lindorm.eu.auth0.com/.well-known/jwks.json",
    "mfaChallengeEndpoint": "https://lindorm.eu.auth0.com/mfa/challenge",
    "registrationEndpoint": "https://lindorm.eu.auth0.com/oidc/register",
    "requestParameterSupported": false,
    "requestUriParameterSupported": false,
    "responseModesSupported": [
      "query",
      "fragment",
      "form_post",
    ],
    "responseTypesSupported": [
      "code",
      "token",
      "id_token",
      "code token",
      "code id_token",
      "token id_token",
      "code token id_token",
    ],
    "revocationEndpoint": "https://lindorm.eu.auth0.com/oauth/revoke",
    "scopesSupported": [
      "openid",
      "profile",
      "offline_access",
      "name",
      "given_name",
      "family_name",
      "nickname",
      "email",
      "email_verified",
      "picture",
      "created_at",
      "identities",
      "phone",
      "address",
    ],
    "subjectTypesSupported": [
      "public",
    ],
    "tokenEndpoint": "https://lindorm.eu.auth0.com/oauth/token",
    "tokenEndpointAuthMethodsSupported": [
      "client_secret_basic",
      "client_secret_post",
      "private_key_jwt",
    ],
    "tokenEndpointAuthSigningAlgValuesSupported": [
      "RS256",
      "RS384",
      "PS256",
    ],
    "userinfoEndpoint": "https://lindorm.eu.auth0.com/userinfo",
  },
  {
    "authorizationEndpoint": "https://lindorm.eu.auth0.com/authorize",
    "claimsSupported": [
      "aud",
      "auth_time",
      "created_at",
      "email",
      "email_verified",
      "exp",
      "family_name",
      "given_name",
      "iat",
      "identities",
      "iss",
      "name",
      "nickname",
      "phone_number",
      "picture",
      "sub",
    ],
    "codeChallengeMethodsSupported": [
      "S256",
      "plain",
    ],
    "deviceAuthorizationEndpoint": "https://lindorm.eu.auth0.com/oauth/device/code",
    "idTokenSigningAlgValuesSupported": [
      "HS256",
      "RS256",
      "PS256",
    ],
    "issuer": "https://lindorm.jp.auth0.io/",
    "jwksUri": "https://lindorm.jp.auth0.com/.well-known/jwks.json",
    "mfaChallengeEndpoint": "https://lindorm.eu.auth0.com/mfa/challenge",
    "registrationEndpoint": "https://lindorm.eu.auth0.com/oidc/register",
    "requestParameterSupported": false,
    "requestUriParameterSupported": false,
    "responseModesSupported": [
      "query",
      "fragment",
      "form_post",
    ],
    "responseTypesSupported": [
      "code",
      "token",
      "id_token",
      "code token",
      "code id_token",
      "token id_token",
      "code token id_token",
    ],
    "revocationEndpoint": "https://lindorm.eu.auth0.com/oauth/revoke",
    "scopesSupported": [
      "openid",
      "profile",
      "offline_access",
      "name",
      "given_name",
      "family_name",
      "nickname",
      "email",
      "email_verified",
      "picture",
      "created_at",
      "identities",
      "phone",
      "address",
    ],
    "subjectTypesSupported": [
      "public",
    ],
    "tokenEndpoint": "https://lindorm.eu.auth0.com/oauth/token",
    "tokenEndpointAuthMethodsSupported": [
      "client_secret_basic",
      "client_secret_post",
      "private_key_jwt",
    ],
    "tokenEndpointAuthSigningAlgValuesSupported": [
      "RS256",
      "RS384",
      "PS256",
    ],
    "userinfoEndpoint": "https://lindorm.eu.auth0.com/userinfo",
  },
]
`;
